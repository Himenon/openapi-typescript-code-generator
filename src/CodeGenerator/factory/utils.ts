import { EOL } from "os";

export interface Comment {
  hasTrailingNewLine: boolean;
  value: string;
}

export const isOnlyAlphabetText = (text: string): boolean => {
  return new RegExp(/^[a-zA-Z]+$/).test(text);
};

export const escapeIdentiferText = (text: string): string => {
  return text.replace(/-/g, "_");
};

export const generateComment = (comment: string, deprecated?: boolean): Comment => {
  const splitComments = deprecated ? ["@deprecated"].concat(comment.split(EOL)) : comment.split(EOL);
  const comments = splitComments.filter((comment, index) => {
    if (index === splitComments.length - 1 && comment === "") {
      return false;
    }
    return true;
  });
  if (comments.length === 1) {
    return {
      hasTrailingNewLine: true,
      value: "* " + comments.join("") + " ",
    };
  }
  return {
    hasTrailingNewLine: true,
    value: "*" + EOL + comments.map(comment => ` * ${comment}`).join(EOL) + EOL + " ",
  };
};
