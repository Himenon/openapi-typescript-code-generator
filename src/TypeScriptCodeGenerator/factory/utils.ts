import { EOL } from "os";

export const generateComment = (comment: string, deprecated?: boolean): string => {
  const splitComments = deprecated ? ["@deprecated"].concat(comment.split(EOL)) : comment.split(EOL);
  const comments = splitComments.filter((comment, index) => {
    if (index === splitComments.length - 1 && comment === "") {
      return false;
    }
    return true;
  });
  return "*" + EOL + comments.map(comment => ` * ${comment}`).join(EOL) + EOL + " ";
};
