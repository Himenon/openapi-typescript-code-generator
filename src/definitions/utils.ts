import { JSONSchema7 } from "json-schema";

export type JSONSchema = JSONSchema7;

export type MapLike<K extends string, T> = {
  [key in K]: T;
};
