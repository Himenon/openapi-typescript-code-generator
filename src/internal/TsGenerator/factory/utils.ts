import { EOL } from "os";

export interface Comment {
  hasTrailingNewLine: boolean;
  value: string;
}

export const isOnlyAlphabetText = (text: string): boolean => {
  return new RegExp(/^[a-zA-Z]+$/).test(text);
};

export const escapeIdentiferText = (text: string): string => {
  return text.replace(/-/g, "_");
};

export const generateComment = (comment: string, deprecated?: boolean): Comment => {
  const excapedComment = comment
    .replace(/\*\//, "\\*\\\\/") // */  -> \*\/
    .replace(/\/\*/, "/\\\\*") // /*  -> \/\*
    .replace(/\*\/\*/, "\\*\\/\\*"); // */* -> \*\/\*
  const splitComments = deprecated ? ["@deprecated"].concat(excapedComment.split(/\r?\n/)) : excapedComment.split(/\r?\n/);
  const comments = splitComments.filter((comment, index) => {
    if (index === splitComments.length - 1 && comment === "") {
      return false;
    }
    return true;
  });
  if (comments.length === 1) {
    return {
      hasTrailingNewLine: true,
      value: "* " + comments.join("") + " ",
    };
  }
  return {
    hasTrailingNewLine: true,
    value: "*" + EOL + comments.map(comment => ` * ${comment}`).join(EOL) + EOL + " ",
  };
};
