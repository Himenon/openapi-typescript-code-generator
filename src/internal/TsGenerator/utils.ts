import ts from "typescript";

import type { CodeGenerator } from "../../types";

export const stringToStatements = (code: string): ts.Statement[] => {
  const source = ts.createSourceFile("", code, ts.ScriptTarget.ESNext, false, ts.ScriptKind.TS);
  return Array.from(source.statements);
};

export const convertIntermediateCodes = (intermediateCodes: CodeGenerator.IntermediateCode[]): ts.Statement[] => {
  return intermediateCodes.reduce<ts.Statement[]>((result, intermediateCode) => {
    if (typeof intermediateCode === "string") {
      return [...result, ...stringToStatements(intermediateCode)];
    }
    return result.concat(intermediateCode);
  }, []);
};
