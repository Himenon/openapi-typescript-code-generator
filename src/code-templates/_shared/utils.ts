import ts from "typescript";

import type { TsGenerator } from "../../api";
import * as Utils from "../../utils";

export interface StringItem {
  type: "string";
  value: string;
}

export interface ExpressionItem {
  type: "property";
  value: ts.Expression;
}

export type Item = StringItem | ExpressionItem;

export type Params$TemplateExpression = Item[];

const getTemplateSpan = (
  factory: TsGenerator.Factory.Type,
  currentIndex: number,
  nextIndex: number,
  lastIndex: number,
  currentItem: ExpressionItem,
  nextItem: Item | undefined,
): ts.TemplateSpan[] => {
  // == last
  if (!nextItem) {
    return [
      factory.TemplateSpan.create({
        expression: currentItem.value,
        literal: factory.TemplateTail.create({
          text: "",
        }),
      }),
    ];
  }

  if (nextItem.type === "property") {
    if (lastIndex === nextIndex) {
      return [
        factory.TemplateSpan.create({
          expression: currentItem.value,
          literal: factory.TemplateTail.create({
            text: "",
          }),
        }),
        factory.TemplateSpan.create({
          expression: nextItem.value,
          literal: factory.TemplateTail.create({
            text: "",
          }),
        }),
      ];
    } else {
      return [
        factory.TemplateSpan.create({
          expression: currentItem.value,
          literal: factory.TemplateTail.create({
            text: "",
          }),
        }),
        factory.TemplateSpan.create({
          expression: nextItem.value,
          literal: factory.TemplateMiddle.create({
            text: "",
          }),
        }),
      ];
    }
  } else {
    if (lastIndex === nextIndex) {
      return [
        factory.TemplateSpan.create({
          expression: currentItem.value,
          literal: factory.TemplateTail.create({
            text: nextItem.value,
          }),
        }),
      ];
    } else {
      return [
        factory.TemplateSpan.create({
          expression: currentItem.value,
          literal: factory.TemplateMiddle.create({
            text: nextItem.value,
          }),
        }),
      ];
    }
  }
};

/**
 *
 * ``
 * `a`
 * `${b}`
 * `a${b}`
 * `${a}${b}`
 * `${a}b${c}`
 * ``
 */
export const generateTemplateExpression = (factory: TsGenerator.Factory.Type, list: Params$TemplateExpression): ts.Expression => {
  if (list.length === 0) {
    return factory.NoSubstitutionTemplateLiteral.create({
      text: "",
    });
  }
  const templateHead = factory.TemplateHead.create({
    text: list[0].type === "property" ? "" : list[0].value,
  });

  const spanList = list.splice(1, list.length);
  if (spanList.length === 0 && list[0].type === "string") {
    return factory.NoSubstitutionTemplateLiteral.create({
      text: list[0].value,
    });
  }
  const lastIndex = spanList.length - 1;
  const restValue = lastIndex % 2;
  let templateSpans: ts.TemplateSpan[] = [];
  for (let i = 0; i <= (lastIndex - restValue) / 2; i++) {
    if (spanList.length === 0) {
      continue;
    }
    const currentIndex = 2 * i;
    const nextIndex = 2 * i + 1;
    const currentItem = spanList[currentIndex];
    const nextItem = spanList[nextIndex];
    if (currentItem.type === "string") {
      throw new Error("Logic Error");
    }
    templateSpans = templateSpans.concat(getTemplateSpan(factory, currentIndex, nextIndex, lastIndex, currentItem, nextItem));
  }
  return factory.TemplateExpression.create({
    head: templateHead,
    templateSpans: templateSpans,
  });
};

export interface VariableAccessIdentifer {
  kind: "string";
  value: string;
}

export interface ElementAccessExpression {
  kind: "element-access";
  value: string;
}

export type VariableElement = VariableAccessIdentifer | ElementAccessExpression;

export const splitVariableText = (text: string): VariableElement[] => {
  // ["...."] にマッチする
  const pattern = '["[a-zA-Z_0-9.]+"]';
  // 'a.b.c["a"]["b"]'.split(/(\["[a-zA-Z_0-9\.]+"\])/g)
  const splitTexts = text.split(/(\["[a-zA-Z_0-9\\.]+"\])/g); // 区切り文字も含めて分割
  return splitTexts.reduce<VariableElement[]>((splitList, value) => {
    if (value === "") {
      return splitList;
    }
    // ["book.name"] にマッチするか
    if (new RegExp(pattern).test(value)) {
      // ["book.name"] から book.name を抽出
      const matchedValue = value.match(/[a-zA-Z_0-9\\.]+/);
      if (matchedValue) {
        splitList.push({
          kind: "element-access",
          value: matchedValue[0],
        });
      }
      return splitList;
    } else {
      const dotSplited = value.split(".");
      const items = dotSplited.map<VariableAccessIdentifer>(childValue => ({ kind: "string", value: childValue }));
      return splitList.concat(items);
    }
  }, []);
};

export const generateVariableIdentifier = (
  factory: TsGenerator.Factory.Type,
  name: string,
): ts.Identifier | ts.PropertyAccessExpression | ts.ElementAccessExpression => {
  if (name.startsWith("/")) {
    throw new Error("can't start '/'. name=" + name);
  }
  const list = splitVariableText(name);
  // Object参照していない変数名の場合
  if (list.length === 1) {
    return factory.Identifier.create({
      name: name,
    });
  }
  const [n1, n2, ...rest] = list;
  // a.b のような単純な変数名の場合
  const first = factory.PropertyAccessExpression.create({
    expression: n1.value,
    name: n2.value,
  });

  return rest.reduce<ts.PropertyAccessExpression | ts.ElementAccessExpression>((previous, current) => {
    if (current.kind === "string" && Utils.isAvailableVariableName(current.value)) {
      return factory.PropertyAccessExpression.create({
        expression: previous,
        name: current.value,
      });
    }
    // 直接 .value でアクセスできない場合に ["value"] といった形で参照する
    return factory.ElementAccessExpression.create({
      expression: previous,
      index: current.value,
    });
  }, first);
};

export interface LiteralExpressionObject {
  [key: string]: { type: "string" | "variable"; value: string };
}

export const generateObjectLiteralExpression = (
  factory: TsGenerator.Factory.Type,
  obj: LiteralExpressionObject,
  extraProperties: ts.PropertyAssignment[] = [],
): ts.ObjectLiteralExpression => {
  const properties = Object.entries(obj).map(([key, item]) => {
    const initializer =
      item.type === "variable" ? generateVariableIdentifier(factory, item.value) : factory.StringLiteral.create({ text: item.value });
    return factory.PropertyAssignment.create({
      name: Utils.escapeText(key),
      initializer,
    });
  });

  return factory.ObjectLiteralExpression.create({
    properties: extraProperties.concat(properties),
    multiLine: true,
  });
};

/**
 * "/{a}/b/{a}/c{a}/".split(new RegExp("({a})"))
 * => ["/", "{a}", "/b/", "{a}", "/c", "{a}", "/"]
 */
export const stringToArray = (text: string, delimiter: string): string[] => text.split(new RegExp(`(${delimiter})`));

export const multiSplitStringToArray = (text: string, delimiters: string[]): string[] => {
  return delimiters.reduce<string[]>(
    (current, delimiter) => {
      return current.reduce<string[]>((result, currentText) => {
        return result.concat(stringToArray(currentText, delimiter));
      }, []);
    },
    [text],
  );
};
