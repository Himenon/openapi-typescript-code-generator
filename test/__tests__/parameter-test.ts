import * as fs from "fs";
import * as path from "path";

import * as Utils from "../utils";

describe("Parameter", () => {
  test("api.test.domain", () => {
    const generateCode = fs.readFileSync(path.join(__dirname, "../code/parameter/api.test.domain.json"), { encoding: "utf-8" });
    const text = Utils.replaceVersionInfo(generateCode);
    expect(text).toMatchSnapshot();
  });
  test("infer.domain", () => {
    const generateCode = fs.readFileSync(path.join(__dirname, "../code/parameter/infer.domain.json"), { encoding: "utf-8" });
    const text = Utils.replaceVersionInfo(generateCode);
    expect(text).toMatchSnapshot();
  });
});
