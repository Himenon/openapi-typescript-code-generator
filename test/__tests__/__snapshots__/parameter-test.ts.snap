// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parameter api.test.domain 1`] = `
"[
  {
    \\"operationId\\": \\"getIncludeLocalReference\\",
    \\"convertedParams\\": {
      \\"escapedOperationId\\": \\"getIncludeLocalReference\\",
      \\"argumentParamsTypeDeclaration\\": \\"Params$getIncludeLocalReference\\",
      \\"functionName\\": \\"getIncludeLocalReference\\",
      \\"requestContentTypeName\\": \\"RequestContentType$getIncludeLocalReference\\",
      \\"responseContentTypeName\\": \\"ResponseContentType$getIncludeLocalReference\\",
      \\"parameterName\\": \\"Parameter$getIncludeLocalReference\\",
      \\"requestBodyName\\": \\"RequestBody$getIncludeLocalReference\\",
      \\"hasRequestBody\\": false,
      \\"hasParameter\\": true,
      \\"pickedParameters\\": [
        {
          \\"name\\": \\"StringQuery\\",
          \\"in\\": \\"query\\",
          \\"required\\": true
        }
      ],
      \\"requestContentTypes\\": [],
      \\"responseSuccessNames\\": [
        \\"Response$getIncludeLocalReference$Status$200\\"
      ],
      \\"responseFirstSuccessName\\": \\"Response$getIncludeLocalReference$Status$200\\",
      \\"has2OrMoreSuccessNames\\": false,
      \\"responseErrorNames\\": [],
      \\"has2OrMoreRequestContentTypes\\": false,
      \\"successResponseContentTypes\\": [
        \\"application/json\\"
      ],
      \\"successResponseFirstContentType\\": \\"application/json\\",
      \\"has2OrMoreSuccessResponseContentTypes\\": false,
      \\"hasAdditionalHeaders\\": false,
      \\"hasQueryParameters\\": true
    },
    \\"operationParams\\": {
      \\"httpMethod\\": \\"get\\",
      \\"requestUri\\": \\"/get/IncludeLocalReference\\",
      \\"comment\\": \\"\\",
      \\"deprecated\\": false,
      \\"parameters\\": [
        {
          \\"name\\": \\"StringQuery\\",
          \\"in\\": \\"query\\",
          \\"description\\": \\"parameters.StringQueryParams\\",
          \\"required\\": true,
          \\"schema\\": {
            \\"type\\": \\"string\\"
          }
        }
      ],
      \\"responses\\": {
        \\"200\\": {
          \\"description\\": \\"Success response of getA\\",
          \\"content\\": {
            \\"application/json\\": {
              \\"schema\\": {
                \\"type\\": \\"object\\",
                \\"required\\": [
                  \\"meta\\"
                ],
                \\"properties\\": {
                  \\"meta\\": {
                    \\"type\\": \\"string\\"
                  }
                }
              }
            }
          }
        }
      }
    }
  },
  {
    \\"operationId\\": \\"getIncludeRemoteReference\\",
    \\"convertedParams\\": {
      \\"escapedOperationId\\": \\"getIncludeRemoteReference\\",
      \\"argumentParamsTypeDeclaration\\": \\"Params$getIncludeRemoteReference\\",
      \\"functionName\\": \\"getIncludeRemoteReference\\",
      \\"requestContentTypeName\\": \\"RequestContentType$getIncludeRemoteReference\\",
      \\"responseContentTypeName\\": \\"ResponseContentType$getIncludeRemoteReference\\",
      \\"parameterName\\": \\"Parameter$getIncludeRemoteReference\\",
      \\"requestBodyName\\": \\"RequestBody$getIncludeRemoteReference\\",
      \\"hasRequestBody\\": true,
      \\"hasParameter\\": true,
      \\"pickedParameters\\": [
        {
          \\"name\\": \\"IncludeRemoteReference\\",
          \\"in\\": \\"query\\",
          \\"required\\": true
        }
      ],
      \\"requestContentTypes\\": [
        \\"application/json\\"
      ],
      \\"requestFirstContentType\\": \\"application/json\\",
      \\"responseSuccessNames\\": [],
      \\"has2OrMoreSuccessNames\\": false,
      \\"responseErrorNames\\": [],
      \\"has2OrMoreRequestContentTypes\\": false,
      \\"successResponseContentTypes\\": [],
      \\"has2OrMoreSuccessResponseContentTypes\\": false,
      \\"hasAdditionalHeaders\\": false,
      \\"hasQueryParameters\\": true
    },
    \\"operationParams\\": {
      \\"httpMethod\\": \\"get\\",
      \\"requestUri\\": \\"/get/IncludeRemoteReference\\",
      \\"comment\\": \\"\\",
      \\"deprecated\\": false,
      \\"requestBody\\": {
        \\"description\\": \\"Request body string type\\",
        \\"content\\": {
          \\"application/json\\": {
            \\"schema\\": {
              \\"type\\": \\"string\\"
            }
          }
        }
      },
      \\"parameters\\": [
        {
          \\"name\\": \\"IncludeRemoteReference\\",
          \\"in\\": \\"query\\",
          \\"description\\": \\"remote reference parameter\\",
          \\"required\\": true,
          \\"schema\\": {
            \\"type\\": \\"number\\"
          }
        }
      ],
      \\"responses\\": {
        \\"200\\": {
          \\"description\\": \\"OK\\"
        }
      }
    }
  },
  {
    \\"operationId\\": \\"getFullRemoteReference\\",
    \\"convertedParams\\": {
      \\"escapedOperationId\\": \\"getFullRemoteReference\\",
      \\"argumentParamsTypeDeclaration\\": \\"Params$getFullRemoteReference\\",
      \\"functionName\\": \\"getFullRemoteReference\\",
      \\"requestContentTypeName\\": \\"RequestContentType$getFullRemoteReference\\",
      \\"responseContentTypeName\\": \\"ResponseContentType$getFullRemoteReference\\",
      \\"parameterName\\": \\"Parameter$getFullRemoteReference\\",
      \\"requestBodyName\\": \\"RequestBody$getFullRemoteReference\\",
      \\"hasRequestBody\\": false,
      \\"hasParameter\\": true,
      \\"pickedParameters\\": [
        {
          \\"name\\": \\"FullRemoteReferenceQuery\\",
          \\"in\\": \\"query\\",
          \\"required\\": true
        }
      ],
      \\"requestContentTypes\\": [],
      \\"responseSuccessNames\\": [
        \\"Response$getFullRemoteReference$Status$200\\"
      ],
      \\"responseFirstSuccessName\\": \\"Response$getFullRemoteReference$Status$200\\",
      \\"has2OrMoreSuccessNames\\": false,
      \\"responseErrorNames\\": [],
      \\"has2OrMoreRequestContentTypes\\": false,
      \\"successResponseContentTypes\\": [
        \\"application/json\\"
      ],
      \\"successResponseFirstContentType\\": \\"application/json\\",
      \\"has2OrMoreSuccessResponseContentTypes\\": false,
      \\"hasAdditionalHeaders\\": false,
      \\"hasQueryParameters\\": true
    },
    \\"operationParams\\": {
      \\"httpMethod\\": \\"get\\",
      \\"requestUri\\": \\"/FullRemoteReference\\",
      \\"comment\\": \\"\\",
      \\"deprecated\\": false,
      \\"parameters\\": [
        {
          \\"name\\": \\"FullRemoteReferenceQuery\\",
          \\"in\\": \\"query\\",
          \\"description\\": \\"Full Remote Reference\\",
          \\"required\\": true,
          \\"schema\\": {
            \\"type\\": \\"string\\",
            \\"description\\": \\"for parameters\\"
          }
        }
      ],
      \\"responses\\": {
        \\"200\\": {
          \\"description\\": \\"response A definition\\",
          \\"headers\\": {
            \\"HeaderA\\": {
              \\"description\\": \\"Header A\\",
              \\"schema\\": {
                \\"type\\": \\"integer\\"
              }
            },
            \\"HeaderB\\": {
              \\"description\\": \\"HeaderB\\",
              \\"schema\\": {
                \\"type\\": \\"string\\"
              }
            }
          },
          \\"content\\": {
            \\"application/json\\": {
              \\"schema\\": {
                \\"type\\": \\"object\\",
                \\"properties\\": {
                  \\"name\\": {
                    \\"type\\": \\"string\\",
                    \\"enum\\": [
                      \\"responseA\\"
                    ],
                    \\"description\\": \\"responseA description\\"
                  }
                }
              }
            }
          }
        }
      }
    }
  },
  {
    \\"operationId\\": \\"i.have.dot\\",
    \\"convertedParams\\": {
      \\"escapedOperationId\\": \\"i$have$dot\\",
      \\"argumentParamsTypeDeclaration\\": \\"Params$i$have$dot\\",
      \\"functionName\\": \\"i$have$dot\\",
      \\"requestContentTypeName\\": \\"RequestContentType$i$have$dot\\",
      \\"responseContentTypeName\\": \\"ResponseContentType$i$have$dot\\",
      \\"parameterName\\": \\"Parameter$i$have$dot\\",
      \\"requestBodyName\\": \\"RequestBody$i$have$dot\\",
      \\"hasRequestBody\\": false,
      \\"hasParameter\\": false,
      \\"pickedParameters\\": [],
      \\"requestContentTypes\\": [],
      \\"responseSuccessNames\\": [
        \\"Response$i$have$dot$Status$200\\"
      ],
      \\"responseFirstSuccessName\\": \\"Response$i$have$dot$Status$200\\",
      \\"has2OrMoreSuccessNames\\": false,
      \\"responseErrorNames\\": [],
      \\"has2OrMoreRequestContentTypes\\": false,
      \\"successResponseContentTypes\\": [
        \\"application/json\\"
      ],
      \\"successResponseFirstContentType\\": \\"application/json\\",
      \\"has2OrMoreSuccessResponseContentTypes\\": false,
      \\"hasAdditionalHeaders\\": false,
      \\"hasQueryParameters\\": false
    },
    \\"operationParams\\": {
      \\"httpMethod\\": \\"get\\",
      \\"requestUri\\": \\"/i/have/dot\\",
      \\"comment\\": \\"\\",
      \\"deprecated\\": false,
      \\"parameters\\": [],
      \\"responses\\": {
        \\"200\\": {
          \\"description\\": \\"Get Books\\",
          \\"content\\": {
            \\"application/json\\": {
              \\"schema\\": {
                \\"type\\": \\"object\\",
                \\"properties\\": {
                  \\"books\\": {
                    \\"type\\": \\"array\\",
                    \\"items\\": {
                      \\"type\\": \\"object\\",
                      \\"description\\": \\"Item Model\\",
                      \\"required\\": [
                        \\"name\\",
                        \\"children\\"
                      ],
                      \\"properties\\": {
                        \\"name\\": {
                          \\"type\\": \\"string\\"
                        },
                        \\"children\\": {
                          \\"type\\": \\"array\\",
                          \\"items\\": {
                            \\"type\\": \\"object\\",
                            \\"required\\": [
                              \\"id\\",
                              \\"name\\"
                            ],
                            \\"properties\\": {
                              \\"id\\": {
                                \\"type\\": \\"string\\"
                              },
                              \\"name\\": {
                                \\"type\\": \\"string\\",
                                \\"description\\": \\"child name\\"
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  },
  {
    \\"operationId\\": \\"getReferenceItems\\",
    \\"convertedParams\\": {
      \\"escapedOperationId\\": \\"getReferenceItems\\",
      \\"argumentParamsTypeDeclaration\\": \\"Params$getReferenceItems\\",
      \\"functionName\\": \\"getReferenceItems\\",
      \\"requestContentTypeName\\": \\"RequestContentType$getReferenceItems\\",
      \\"responseContentTypeName\\": \\"ResponseContentType$getReferenceItems\\",
      \\"parameterName\\": \\"Parameter$getReferenceItems\\",
      \\"requestBodyName\\": \\"RequestBody$getReferenceItems\\",
      \\"hasRequestBody\\": false,
      \\"hasParameter\\": false,
      \\"pickedParameters\\": [],
      \\"requestContentTypes\\": [],
      \\"responseSuccessNames\\": [
        \\"Response$getReferenceItems$Status$200\\"
      ],
      \\"responseFirstSuccessName\\": \\"Response$getReferenceItems$Status$200\\",
      \\"has2OrMoreSuccessNames\\": false,
      \\"responseErrorNames\\": [],
      \\"has2OrMoreRequestContentTypes\\": false,
      \\"successResponseContentTypes\\": [
        \\"application/json\\"
      ],
      \\"successResponseFirstContentType\\": \\"application/json\\",
      \\"has2OrMoreSuccessResponseContentTypes\\": false,
      \\"hasAdditionalHeaders\\": false,
      \\"hasQueryParameters\\": false
    },
    \\"operationParams\\": {
      \\"httpMethod\\": \\"get\\",
      \\"requestUri\\": \\"/get/reference/items\\",
      \\"comment\\": \\"\\",
      \\"deprecated\\": false,
      \\"parameters\\": [],
      \\"responses\\": {
        \\"200\\": {
          \\"description\\": \\"Get Books\\",
          \\"content\\": {
            \\"application/json\\": {
              \\"schema\\": {
                \\"type\\": \\"object\\",
                \\"properties\\": {
                  \\"books\\": {
                    \\"type\\": \\"array\\",
                    \\"items\\": {
                      \\"type\\": \\"object\\",
                      \\"description\\": \\"Item Model\\",
                      \\"required\\": [
                        \\"name\\",
                        \\"children\\"
                      ],
                      \\"properties\\": {
                        \\"name\\": {
                          \\"type\\": \\"string\\"
                        },
                        \\"children\\": {
                          \\"type\\": \\"array\\",
                          \\"items\\": {
                            \\"type\\": \\"object\\",
                            \\"required\\": [
                              \\"id\\",
                              \\"name\\"
                            ],
                            \\"properties\\": {
                              \\"id\\": {
                                \\"type\\": \\"string\\"
                              },
                              \\"name\\": {
                                \\"type\\": \\"string\\",
                                \\"description\\": \\"child name\\"
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  },
  {
    \\"operationId\\": \\"searchBook\\",
    \\"convertedParams\\": {
      \\"escapedOperationId\\": \\"searchBook\\",
      \\"argumentParamsTypeDeclaration\\": \\"Params$searchBook\\",
      \\"functionName\\": \\"searchBook\\",
      \\"requestContentTypeName\\": \\"RequestContentType$searchBook\\",
      \\"responseContentTypeName\\": \\"ResponseContentType$searchBook\\",
      \\"parameterName\\": \\"Parameter$searchBook\\",
      \\"requestBodyName\\": \\"RequestBody$searchBook\\",
      \\"hasRequestBody\\": false,
      \\"hasParameter\\": true,
      \\"pickedParameters\\": [
        {
          \\"name\\": \\"book.name\\",
          \\"in\\": \\"path\\",
          \\"required\\": true
        },
        {
          \\"name\\": \\"from.publishedAt\\",
          \\"in\\": \\"query\\"
        }
      ],
      \\"requestContentTypes\\": [],
      \\"responseSuccessNames\\": [
        \\"Response$searchBook$Status$200\\"
      ],
      \\"responseFirstSuccessName\\": \\"Response$searchBook$Status$200\\",
      \\"has2OrMoreSuccessNames\\": false,
      \\"responseErrorNames\\": [],
      \\"has2OrMoreRequestContentTypes\\": false,
      \\"successResponseContentTypes\\": [
        \\"application/json\\"
      ],
      \\"successResponseFirstContentType\\": \\"application/json\\",
      \\"has2OrMoreSuccessResponseContentTypes\\": false,
      \\"hasAdditionalHeaders\\": false,
      \\"hasQueryParameters\\": true
    },
    \\"operationParams\\": {
      \\"httpMethod\\": \\"get\\",
      \\"requestUri\\": \\"/get/search/{book.name}\\",
      \\"comment\\": \\"\\",
      \\"deprecated\\": false,
      \\"parameters\\": [
        {
          \\"in\\": \\"path\\",
          \\"name\\": \\"book.name\\",
          \\"required\\": true,
          \\"schema\\": {
            \\"type\\": \\"string\\"
          }
        },
        {
          \\"in\\": \\"query\\",
          \\"name\\": \\"from.publishedAt\\",
          \\"schema\\": {
            \\"type\\": \\"integer\\",
            \\"format\\": \\"int32\\"
          }
        }
      ],
      \\"responses\\": {
        \\"200\\": {
          \\"description\\": \\"Search Book Result\\",
          \\"content\\": {
            \\"application/json\\": {
              \\"schema\\": {
                \\"type\\": \\"object\\",
                \\"properties\\": {
                  \\"id\\": {
                    \\"type\\": \\"number\\"
                  },
                  \\"bookTitle\\": {
                    \\"type\\": \\"string\\"
                  }
                }
              }
            }
          }
        }
      }
    }
  },
  {
    \\"operationId\\": \\"getBookById\\",
    \\"convertedParams\\": {
      \\"escapedOperationId\\": \\"getBookById\\",
      \\"argumentParamsTypeDeclaration\\": \\"Params$getBookById\\",
      \\"functionName\\": \\"getBookById\\",
      \\"requestContentTypeName\\": \\"RequestContentType$getBookById\\",
      \\"responseContentTypeName\\": \\"ResponseContentType$getBookById\\",
      \\"parameterName\\": \\"Parameter$getBookById\\",
      \\"requestBodyName\\": \\"RequestBody$getBookById\\",
      \\"hasRequestBody\\": false,
      \\"hasParameter\\": true,
      \\"pickedParameters\\": [
        {
          \\"name\\": \\"id\\",
          \\"in\\": \\"path\\",
          \\"required\\": true
        }
      ],
      \\"requestContentTypes\\": [],
      \\"responseSuccessNames\\": [
        \\"Response$getBookById$Status$200\\"
      ],
      \\"responseFirstSuccessName\\": \\"Response$getBookById$Status$200\\",
      \\"has2OrMoreSuccessNames\\": false,
      \\"responseErrorNames\\": [],
      \\"has2OrMoreRequestContentTypes\\": false,
      \\"successResponseContentTypes\\": [
        \\"application/json\\"
      ],
      \\"successResponseFirstContentType\\": \\"application/json\\",
      \\"has2OrMoreSuccessResponseContentTypes\\": false,
      \\"hasAdditionalHeaders\\": false,
      \\"hasQueryParameters\\": false
    },
    \\"operationParams\\": {
      \\"httpMethod\\": \\"get\\",
      \\"requestUri\\": \\"/get/books/{id}\\",
      \\"comment\\": \\"\\",
      \\"deprecated\\": false,
      \\"parameters\\": [
        {
          \\"name\\": \\"id\\",
          \\"in\\": \\"path\\",
          \\"required\\": true,
          \\"description\\": \\"Book ID\\",
          \\"schema\\": {
            \\"type\\": \\"string\\",
            \\"format\\": \\"uuid\\"
          }
        }
      ],
      \\"responses\\": {
        \\"200\\": {
          \\"description\\": \\"get books\\",
          \\"content\\": {
            \\"application/json\\": {
              \\"schema\\": {
                \\"type\\": \\"object\\",
                \\"properties\\": {
                  \\"stringType\\": {
                    \\"type\\": \\"string\\"
                  },
                  \\"numberType\\": {
                    \\"type\\": \\"number\\"
                  },
                  \\"booleanType\\": {
                    \\"type\\": \\"boolean\\"
                  },
                  \\"arrayType\\": {
                    \\"type\\": \\"string\\"
                  },
                  \\"objectType\\": {
                    \\"type\\": \\"object\\",
                    \\"required\\": [
                      \\"requiredStringType\\"
                    ],
                    \\"properties\\": {
                      \\"requiredStringType\\": {
                        \\"type\\": \\"string\\"
                      }
                    }
                  }
                },
                \\"example\\": {
                  \\"stringType\\": \\"string-type\\",
                  \\"numberType\\": 1,
                  \\"booleanType\\": false,
                  \\"arrayType\\": [],
                  \\"objectType\\": {
                    \\"requiredStringType\\": \\"required-string-type\\"
                  }
                }
              }
            }
          }
        }
      }
    }
  },
  {
    \\"operationId\\": \\"deleteBook\\",
    \\"convertedParams\\": {
      \\"escapedOperationId\\": \\"deleteBook\\",
      \\"argumentParamsTypeDeclaration\\": \\"Params$deleteBook\\",
      \\"functionName\\": \\"deleteBook\\",
      \\"requestContentTypeName\\": \\"RequestContentType$deleteBook\\",
      \\"responseContentTypeName\\": \\"ResponseContentType$deleteBook\\",
      \\"parameterName\\": \\"Parameter$deleteBook\\",
      \\"requestBodyName\\": \\"RequestBody$deleteBook\\",
      \\"hasRequestBody\\": false,
      \\"hasParameter\\": true,
      \\"pickedParameters\\": [
        {
          \\"name\\": \\"id\\",
          \\"in\\": \\"path\\",
          \\"required\\": true
        }
      ],
      \\"requestContentTypes\\": [],
      \\"responseSuccessNames\\": [
        \\"Response$deleteBook$Status$200\\"
      ],
      \\"responseFirstSuccessName\\": \\"Response$deleteBook$Status$200\\",
      \\"has2OrMoreSuccessNames\\": false,
      \\"responseErrorNames\\": [],
      \\"has2OrMoreRequestContentTypes\\": false,
      \\"successResponseContentTypes\\": [
        \\"application/json\\"
      ],
      \\"successResponseFirstContentType\\": \\"application/json\\",
      \\"has2OrMoreSuccessResponseContentTypes\\": false,
      \\"hasAdditionalHeaders\\": false,
      \\"hasQueryParameters\\": false
    },
    \\"operationParams\\": {
      \\"httpMethod\\": \\"delete\\",
      \\"requestUri\\": \\"/get/books/{id}\\",
      \\"comment\\": \\"\\",
      \\"deprecated\\": false,
      \\"parameters\\": [
        {
          \\"name\\": \\"id\\",
          \\"in\\": \\"path\\",
          \\"required\\": true,
          \\"description\\": \\"Number Book ID\\",
          \\"schema\\": {
            \\"type\\": \\"number\\",
            \\"format\\": \\"uuid\\"
          }
        }
      ],
      \\"responses\\": {
        \\"200\\": {
          \\"description\\": \\"get books\\",
          \\"content\\": {
            \\"application/json\\": {
              \\"schema\\": {
                \\"type\\": \\"object\\",
                \\"properties\\": {
                  \\"status\\": {
                    \\"type\\": \\"string\\",
                    \\"enum\\": [
                      \\"ok\\"
                    ]
                  }
                }
              }
            }
          }
        }
      }
    }
  }
]"
`;

exports[`Parameter infer.domain 1`] = `"[]"`;
